### PR TITLE
Refactor karaoke-related code

### PR DESCRIPTION
**Pull request by xIceArcher**

### CODE REVIEW

File: lyricsheets/models/karaoke.py [hunk 1]
```diff
@@ -159,248 +166,217 @@ def middle(self) -> float:
...
 
-    def _calculate_positions(self, resX: int = 1920, resY: int = 1080):
-        if not self.style:
-            raise StyleNotBoundException()
+    @property
+    def height(self) -> float:
+        return self.style.fontSize
 
-        # Horizontal positioning
+    @property
```
Comment (ahuei123456) [inline lyricsheets/models/karaoke.py:276]: Any reason why this isn't cached?

File: lyricsheets/models/karaoke.py [hunk 2]
```diff
@@ -159,248 +166,217 @@ def middle(self) -> float:
...
             ) / 2
-            self._x = self.center
-            # Left aligned
-            self._left = self.style.marginL
-            self._x = self.left
+        else:
+            # Right aligned
+            return self.resX - self.style.marginR - self.width
+
+    @property
```
Comment (ahuei123456) [inline lyricsheets/models/karaoke.py:298]: Same as above.

File: lyricsheets/models/karaoke.py [hunk 3]
```diff
@@ -159,248 +166,217 @@ def middle(self) -> float:
...
+        }:
+            # Center aligned
+            return (
+                self.resY - self.style.marginV * 2 - self.height
+            ) / 2 + self.style.marginV
+        else:
+            # Top aligned
+            return self.style.marginV
+
+    @property
```
Comment (ahuei123456) [inline lyricsheets/models/karaoke.py:320]: Same as above.

File: lyricsheets/models/karaoke.py [hunk 4]
```diff
@@ -159,248 +166,217 @@ def middle(self) -> float:
...
-            self._x = self.center
+            return self.center
         else:
-            # Right aligned
-            self._left = resX - self.style.marginR - self.width
-            self._x = self.right
+            return self.right
 
-        # Vertical positioning
+    @property
```
Comment (ahuei123456) [inline lyricsheets/models/karaoke.py:337]: Same as above.

**Review by ahuei123456 — COMMENTED** [review]
In general, I still disagree with the idea of making a separate RomajiKLine and ENKLine. Only one effect (`default_live_karaoke_effect`) would need to calculate and store the actor transition timings, and it's especially not necessary to make it a part of `KaraokeEffect`. 

I suggested previously that the actor transition timings should be stored in `DefaultEffect`, and `KaraokeEffect` left as-is. `to_romaji_k_events` will calculate and store the actor transition timings in the class, and `to_en_k_events` would retrieve those timings and use it for its tags. I still think that's the cleanest way to do the calculations, and it has the benefit of minimizing the code changes necessary.

Regarding changing the variable names in the K* classes, I can see what you're going for, but the whole reason I named those variables that way was to make porting from Aegisub karaoke templates as simple as possible by keeping the variable names the same.

File: lyricsheets/models/karaoke.py (hunk 1 above)
Comment (xIceArcher) [reply lyricsheets/models/karaoke.py:276]: Mainly the reasoning of what is cached and what isn't is just how much work it takes to calculate something, so things that are cached are:
- Things that require FontScaler
- Things that require iterating through a list of things (e.g. the char offsets for one char is depending on all the preceding chars)

**Comment by xIceArcher** [comment]
> In general, I still disagree with the idea of making a separate RomajiKLine and ENKLine. Only one effect (`default_live_karaoke_effect`) would need to calculate and store the actor transition timings, and it's especially not necessary to make it a part of `KaraokeEffect`.
> 
> I suggested previously that the actor transition timings should be stored in `DefaultEffect`, and `KaraokeEffect` left as-is. `to_romaji_k_events` will calculate and store the actor transition timings in the class, and `to_en_k_events` would retrieve those timings and use it for its tags. I still think that's the cleanest way to do the calculations, and it has the benefit of minimizing the code changes necessary.

The problem with that method is that `to_*_k_events` are no longer pure functions - they now have some form of state associated with it.

In that sense the existence of `DefaultKaraokeEffect` now constrains the caller to only ever call these functions in the exact "correct", order. (e.g., it will no longer be possible to, generate all the romaji lines of all songs in a file before generating all the EN lines, or generating a file _without_ romaji lines (for whatever reason))

A major benefit of the current class organization is that the `Effect` are stateless and only contain pure functions. Rather than the typical OOP classes describing objects, these classes contain methods that are closer to "transformers" that merely convert inputs to outputs directly, without taking into account any state information.

I'd argue that if we're going to make these effect classes stateful, we might as well abandon the current design and come up with something else, since if it's not stateless, there's not much benefit to this design in the first place.

> Regarding changing the variable names in the K* classes, I can see what you're going for, but the whole reason I named those variables that way was to make porting from Aegisub karaoke templates as simple as possible by keeping the variable names the same.

Fair point, I didn't consider that, let me see whether there's a way I can work around that...

**Comment by ahuei123456** [comment]
> The problem with that method is that `to_*_k_events` are no longer pure functions - they now have some form of state associated with it.

I agree. But right now, that's solved by introducing a dependency from `ENKLine` to `RomajiKLine`, which I find is a little excessive. That forces any class that inherits from `KaraokeEffect` to maintain that dependency, while my opinion is that since most other karaoke effects (that would inherit from `KaraokeEffect`) do not _need_ that dependency, the dependency should be introduced in `DefaultLiveKaraokeEffect` and no further up the chain.

> In that sense the existence of `DefaultKaraokeEffect` now constrains the caller to only ever call these functions in the exact "correct", order. (e.g., it will no longer be possible to, generate all the romaji lines of all songs in a file before generating all the EN lines, or generating a file _without_ romaji lines (for whatever reason))

I'm not seeing an issue with this. Generating all the romaji lines first is what `to_events()` already does, and any effect that only wants to generate EN shouldn't be inheriting from `DefaultLiveKaraokeEffect`. With my above proposal, the dependency is moved to `DefaultLiveKaraokeEffect`, and any other effects can inherit from `KaraokeEffect`.

> A major benefit of the current class organization is that the Effect are stateless and only contain pure functions. Rather than the typical OOP classes describing objects, these classes contain methods that are closer to "transformers" that merely convert inputs to outputs directly, without taking into account any state information.
> 
> I'd argue that if we're going to make these effect classes stateful, we might as well abandon the current design and come up with something else, since if it's not stateless, there's not much benefit to this design in the first place.

I agree that the `Effect` classes being stateless is a good thing, and should be kept. However, I see no merit in proposing that the same standard be applied to `DefaultLiveKaraokeEffect`, if doing so requires warping the infrastructure around it to create ugly dependencies too far up the inheritance chain. In the end, `DefaultLiveKaraokeChain` is more of a small script that is an example implementation using the infrastructure provided, and doesn't really _need_ to stay stateless.

**Comment by xIceArcher** [comment]
I see your point, how about something like this?

**Comment by ahuei123456** [comment]
Honestly, as someone who will end up writing the scripts to automate custom effects, I'm still not buying that all this additional infrastructure is necessary for the main reason to avoid keeping `DefaultLiveKaraokeEffect` stateless, when just storing _something_ in the class fixes the issue much more easily. As I brought up previously, I don't really consider `DefaultLiveKaraokeEffect` part of the infrastructure, and don't see why it really needs to stay stateless, even though it makes sense for the rest of the infrastructure to be that way. 

One other thing to keep in mind, which I should've brought up earlier. From the perspective of having written scripts to port over other effects to this infrastructure, `DefaultLiveKaraokeEffect` is a script that needs to be extended on for custom effects for songs in live concert subs. Previously, I did that by keeping all the event generation logic in one method and having any other effects that add on to existing lines just call that method, before adding the extra effects. Does the current implementation of `DefaultLiveKaraokeEffect` facilitate the easy extension of effect generation as well? From what I can tell, a lot of the boilerplate code in it will have to be duplicated for every single effect that wants to add on to existing lines. My proposal is for you to take a look at the current scripts I wrote for songs (Tiny Stars, Wish Song, HAPPY PARTY TRAIN, Koi ni Naritai Aquarium), and port them to work with your current implementation of `DefaultLiveKaraokeEffect`. Once you've experienced porting those scripts to work with the new version of `DefaultLiveKaraokeEffect`, or decide that you want to make some changes to it to make it easier to port, then I won't have any other objections regarding this.

**Comment by xIceArcher** [comment]
Right, agreed that backwards compatibility with existing scripts might be a problem.

Let's set aside our differences regarding how `DefaultLiveKaraokeEffect` is implemented for the moment and focus on that: maintaining backwards compatibility as much as possible.

I have taken a look at the scripts and here are my observations.

The things that haven't changed are:
- `KaraokeEffect` still exists
- The signature of the methods (`to_romaji_k_events` and `to_en_k_events`) in `KaraokeEffect` have not changed, so all existing scripted effects (that do not call `to_default_event`) still satisfy this class and can run without any changes

(I notice that none of the custom scripts actually use `DefaultLiveKaraokeEffect`, so any changes to this class shouldn't matter at all, as you said, `DefaultLiveKaraokeEffect` isn't part of the core architecture)

However, the things that changed and breaks existing scripts (all of them since they all actually do call `to_default_event`) are:
- The `to_default_event` method is now missing (and has been replaced with `to_romaji_default_event` and `to_en_default_event`), no matter, I can just add back `to_default_event` and switch based on `isEN`, but...)
- The `to_default_event` now takes an _additional_ parameter of `switchSylFadeOffsets`, which appears to be the bigger problem

Here's the signature and implementation for `to_default_event`, for reference:
```python
def to_default_event(
    line: KLine,
    actorToStyle: Mapping[str, Sequence[pyass.Tag]],
    switchDuration: timedelta,
    transitionDuration: timedelta,
    switchSylFadeOffsets: Sequence[timedelta],
) -> pyass.Event:
    if line.isEN:
        f = to_default_en_event
    else:
        f = to_default_romaji_event
    return f(line, actorToStyle, switchDuration, transitionDuration, switchSylFadeOffsets)
```

Note that the parameter for `switchSylFadeOffsets` _has_ to be different for Romaji and EN lines (specifically, this needs to be calculated from the Romaji lines, regardless of whether it's a Romaji line or an EN line)

Even disregarding the refactoring changes in this MR, changing how the fade works _requires_ this parameter to exist, since this parameter cannot be derived from `KLine` alone (even before this MR, since there's a dependency between EN and Romaji `KLine`s introduced in MR #19 )

So here are the only three options I can think of to fix scripts that were written pre-MR #19 , even disregarding the refactoring in this MR:
- We remove `switchSylFadeOffsets` and somehow encode the information between Romaji and EN `KLine`s within `KLine` itself
- We keep `switchSylFadeOffsets` and require all callers of `to_default_event` to pass it. All scripts that want to use 
`to_default_event` _must_ pass it in the way `DefaultLiveKaraokeEffect` does (which likely requires the effect that intends to use `to_default_event` to inherit from `DependentKaraokeEffect` instead of `KaraokeEffect`)
- We keep the implementation in MR #19, i.e. change `KaraokeEffect`
- We revert MR #19 

The last option is unacceptable. The third is undesirable as well, as per your earlier comment.

The first option is what I tried at first, and I agree with you that after trying it out, introducing separate classes `RomajiKLine` and `ENKLine` isn't exactly ideal, so I'd rather not pursue that route as well.

So it does seem like we're left with the second option, which is what this MR's current state is at (minus the `to_default_event` method)

What do you propose, since you seem adamantly against this change, even aside from our contentions about `DefaultLiveKaraokeEffect`?

**Comment by xIceArcher** [comment]
To reiterate:
- Scripts that don't use `to_default_event` are _completely fine and will require zero changes whatsoever
- For scripts that use `to_default_event`:
  - If we use the approach in MR #19 (i.e. `KLine` itself encodes the relationship between Romaji and EN) , none of the scripts need to change, since the changes are all done in `KaraokeEffect` and no changes are made to the signatures of anything that inherits from  `KaraokeEffect`
  - If we don't use the approach in MR #19 (and keep Romaji and EN `KLine` objects independent), `to_default_event` _needs_ additional parameter(s) so the EN KLine knows the Romaji KLine's fade offsets (**_regardless_** of the implementation of `DefaultLiveKaraokeEffect`, whether stateless or stateful), and all scripts that use `to_default_event` need to be updated to also calculate and pass this parameter

**Comment by ahuei123456** [comment]
How about this?

- Move all the logic within `to_*_k_events` into a separate method within the class, and change `to_*_k_events` to just be `return [other_method(...) for line...]`.
- Any effect that currently uses `to_default_effect` can just inherit from `DefaultLiveKaraokeEffect` and add a filter by line number to determine what lines need additional logic other than just calling `other_method(...)`.

Or the alternative would be what I proposed earlier. Since `to_romaji_k_events` will always be called _before_ `to_en_k_events`, the first call of `to_default_event` within `to_romaji_k_events` will calculate and save all the fade offsets, which the call of `to_default_event` from `to_en_k_events` will retrieve and use as a tag. This data could even be tagged via start time/line duration so that it could be run concurrently for multiple songs at once. This would mean that the method signature for `to_default_event` _does not change_, so existing scripts do not need to be touched at all.

**Comment by xIceArcher** [comment]
Seems good.

Irrelevant side note:
The alternative proposal now sounds even worse to me since it's not just a stateful method (which is perfectly valid usually though I'd argue not in this situation, anyway I digress), it's now a stateful free-floating function (`to_default_event`), which is even worse since it's basically the same as having global variables so...

**Pull request merged**